### DISCUSSION TITLE
How to call Route Handler within client component?

### DISCUSSION THREAD

**Discussion by skorphil** [vercel/next.js#58896] [App Router]
Hi, i'm developing app with latest *app router* and *route handler* patterns.

How should I fetch data from API for my component? Documentation says I can "Call Route handler from a client" https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-client-with-route-handlers

What does this mean? 

Currently, I have:

*api/route.js* (my Route handler)
```js
export async function GET(request) {
  const { searchParams } = new URL(request.url)
  ...
  return Response.json(data);
}
```
I can access it via `fetch('mysite.com/api...')`
But how can I "call" it from client?

*Component.jsx*
```jsx
import { GET } from 'api/route.js' // ?

export function Component() {
  const data = GET() // ?
  ...
}
```

At first, I thought calling == fetching, but I created client component And since fetch requires async, I made it async:
*MyComponent.jsx*
```js
'use client'

export async function MyComponent() {
  const response = await fetch(`http://MY_ROUTE_HANDLER_ADDRESS`);
  const data = await response.json();
  console.log(data);
```

And I got error:
```
Unhandled Runtime Error
Error: async/await is not yet supported in Client Components, only Server Components. This error is often caused by accidentally adding `'use client'` to a module that was originally written for the server.
```

So, after all how should I "call route handler"? There is no examples in documentation

**Comment by icyJoseph** [comment]
Hi,

You have to use an effect, and state, and later on, the React team says you can use a new hook called `use`, but let's not talk about that right now.

I think the docs are assuming that you understand `fetch` requests, and how these are handled from the Client, independently of React, or any other framework.

For the React specific case though, typically, we couple to a promise in an use-effect:

```js
useEffect(() => {
   somePromise().then(/* do something when it resolves */).catch(/* do something when it rejects */)
}, [/* dependencies */]);
```

So the most basic approach, called, `render-then-fetch` is to, have a `useState` hook, initially empty, `null` or whatever, and then when the promise (be that a fetch request), resolves, you set the resolved JSON or any other data, on that `useState` hook.

Call it from a client, is referring to sending a fetch request. Then it is up to you, how to use that request/response. Do remember that it is not only your React client code that can make requests to your handler. It is available for everyone to reach.

Note that this approach can quickly become overwhelming, as asynchronous work does require paying attention to a bunch of details. You might benefit from using `swr` or `react-query`.

Also, this means that you won't have the data on your first render, because on SSR, you'll just have an empty shell. If you want to have data, then refactor the component, so that you can fetch on the server, **_directly from its source, avoiding Router Handlers_**, and pass that as props to the client components.

**Reply by skorphil** [reply]
Thanks for explanation, it become clearer now. So basically call Route Handler == just use fetch as with any another APIs? 
What bothered me is that React documentation suggests not to `useEffect` and rely on "framework specific fetching mechanisms". That's why I thought, that "calling Route Handler" is not "usual fetch", but some "nextjs specific mechanism".

**Reply by icyJoseph** [reply]
Yeah, that's why if the data layer grows too much you should really go ahead with swr or other like tools.

**Comment by Prathapkarumanchi** [comment]
I think you don't need to use 'use client' in that component and do the basic async and await fetch call it should work.
FYR - check this URL the doc says all about Routes calls and not common fetch inside client component.
https://nextjs.org/docs/app/building-your-application/routing/route-handlers

![image](https://github.com/vercel/next.js/assets/53927098/dc59229d-550c-4d20-9214-e503e6ff132f)

**Reply by icyJoseph** [reply]
Preferably, we should avoid calling API Router handlers as part of a Server Component, or data fetching function such as GetServerSideProps:

- https://nextjs-faq.com/fetch-api-in-rsc
- https://nextjs-faq.com/fetch-api-in-getserversideprops

**Reply by skorphil** [reply]
Thanks, i have read that doc, but still it gave me no clues about "calling Route Handlers".
As far as I understand - Route Handler is just API endpoint. Now I fetch data from that API using `fetch('my handler')`. This is the same approach as fetching any other APIs

**Comment by skorphil** [comment]
> Preferably, we should avoid calling API Router handlers as part of a Server Component, or data fetching function such as GetServerSideProps:
> 
> * https://nextjs-faq.com/fetch-api-in-rsc
> * https://nextjs-faq.com/fetch-api-in-getserversideprops

Thanks, valuable info

**Comment by superfly9** [comment]
it helped me a lot.thanks
